crates/sui-core/src/authority.rs

```diff
@@ -1096,12 +1096,39 @@ impl AuthorityState {
 
         // If commit_certificate returns an error, tx_guard will be dropped and the certificate
         // will be persisted in the log for later recovery.
-        let output_keys: Vec<_> = inner_temporary_store
+        let mut output_keys: Vec<_> = inner_temporary_store
             .written
             .iter()
             .map(|(_, ((id, seq, _), obj, _))| InputKey(*id, (!obj.is_package()).then_some(*seq)))
             .collect();
 
+        // add here deleted shared object Inputkeys to the outputkeys that then get sent to notify_commit
+        let mut waiting_output_keys = Vec::new();
+        let deleted_output_keys = inner_temporary_store
+            .deleted
+            .iter()
+            .filter(|(id, _)| {
+                inner_temporary_store
+                    .objects
+                    .get(id)
+                    .is_some_and(|obj| obj.is_shared())
+            })
+            .map(|(id, (seq, _))| {
+                if let Some(v) = epoch_store
+                    .multi_get_next_shared_object_versions(vec![id].into_iter())
+                    .unwrap()[0]
+                {
+                    for i in u64::from(*seq) + 1..=u64::from(v) {
```

Comment:
This creates an easy opportunity for a DoS attack. The range of sequence numbers could be arbitrarily large due to Lamport versioning.
I don't quite understand why you need to notify all versions in the range, instead of just the deleted version. Could you explain the strategy for the following in more detail?
- When a transaction uses a formerly-deleted shared object, how does that shared object manifest in effects (or does it show up at all)?
- When such a transaction enters TransactionManager, what version would it wait for, and how?

Reply:
That is a good point. The reason why I did it this way is because, according to my testing, it is not guaranteed that we enter this code more than once, ie once per previously enqueued transaction that is waiting on a deleted shared object availability signal. So the case we delete a shared object, we would have other transactions are enqueued waiting for later versions of the deleted shared object, so I signal to all those transactions as well here. To be frank, I and am open to iterating on this approach, but I my feeling is its better not to rely on this code being called multiple times. 

> When a transaction uses a formerly-deleted shared object, how does that shared object manifest in effects (or does it show up at all)? 

It doesn't show up at all

> When such a transaction enters TransactionManager, what version would it wait for, and how?

It waits for the next version that the object would have been incremented to. This is surprisingly difficult to reconstruct here, as when I call the same function used before execution here, it increments to the next version beyond what it resulted in when called before.

Reply:
> So the case we delete a shared object, we would have other  . transactions are enqueued waiting for later versions of the deleted shared object, so I signal to all those transactions as well here.

I think the right way to handle this is to let each cert wake the next one in the chain.